examples/cocoa/CocoaAppDelegate.mm

```diff
@@ -31,6 +31,14 @@
 #define GREEN   @"Green"
 #define RED     @"Red"
 
+#if defined(__APPLE__)
+    #if defined(__clang__)
+        #pragma clang diagnostic ignored "-Wdeprecated-declarations"
+    #elif defined(__GNUC__)
+        #pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+    #endif
+#endif
```

Comment:
I think that the code as written doesn't ignore that warning only for this specific file, but also for any other subsequent one. See [gcc docs](https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Pragmas.html).

I think it should be like this:

```cpp
// Around the beginning of file...
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"

// ...stuff...

// Around the end of the file...
#pragma GCC diagnostic pop
```

Reply:
The GCC documentation assumes the user is aware that anything that has an influence on the compilation process only applies to the current invocation. Since a new GCC process is created for every TU we compile it is impossible for any diagnostic state to carry over to future TUs.

We deliberately ignore after the `#include`s and don't pop the state since we want the deprecation warnings to be ignored for the whole file. You can think of there being an implicit pop at the end of the file (where the stack ceases to exist).

Reply:
@binary1248: sure, it's fine in a source file... but in the PR that pattern is used in a few`.hpp` files, that's why I mentioned it.

Reply:
The deprecation warnings were generated from contents in a header file as well. They would be generated in every TU that includes that header, thus the pragma has to be there too as well. It doesn't look "clean" but this is a case where I think a "dirty" solution is necessary to fix a "dirty" problem.

Reply:
Yes, it's going to be propagated through the whole translation unit but not others. Still, the user warnings can be affected so I think @vittorioromeo has a point.

We probably don't want an unguarded (push/pop) pragma in a **header file** because it could suppress the user warnings. https://godbolt.org/z/hnf963n4W

I'll change and see if any new warning emerges.

Reply:
I wonder, though, is checking for `__clang__` and `__GNUC__` necessary at all @binary1248? I mean unknown pragmas will get discarded, won't they?

Reply:
These are internal headers we are talking about, the only code it affects is ours, and since all warnings will cause building to fail I am OK with disabling Apple deprecation warnings wherever necessary in SFML's internals. The day Apple decides that these APIs aren't just deprecated but also removes them CI will fail to build completely and that will warrant a complete rewrite of the Apple implementation in SFML. Of course, if anybody is willing to take the time to rewrite now, they are free to do so as well.

Regarding unknown pragmas: I have seen instances where compilers warn about them, since we are building with `-Werror` those warnings will cause building to fail.